com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboard.cs

```diff
@@ -23,23 +100,52 @@ class SGBlackboard : GraphSubWindow, ISelection
         public override string UxmlName => "GraphView/Blackboard";
         public override string layoutKey => "UnityEditor.ShaderGraph.Blackboard";
 
-        public Action<SGBlackboard> addItemRequested { get; set; }
-        public Action<SGBlackboard, int, VisualElement> moveItemRequested { get; set; }
-        public Action<SGBlackboard, VisualElement, string> editTextRequested { get; set; }
+        public Action addItemRequested { get; set; }
+        public Action<int, VisualElement> moveItemRequested { get; set; }
+
+        GenericMenu m_AddPropertyMenu;
 
-        public SGBlackboard(GraphView associatedGraphView) : base(associatedGraphView)
+        public SGBlackboard(BlackboardViewModel viewModel) : base(viewModel)
         {
+            ViewModel = viewModel;
```

Comment:
I'm not certain if its a pattern that's been formally described anywhere, but you've got the core idea there, where the view model classes themselves also have a hierarchy, and the base view model interface `ISGViewModel` is what the parent class of `SGBlackboard`, i.e. `GraphSubWindow` is maintaining a reference to

The blackboard subclass (and other views in general), will maintain references to the same view model instance, but at a lower derived class level that is specific to their view model needs, i.e. for the ``SGBlackboard`, we have the `BlackboardViewModel` reference and so on.

One could argue that the base class reference to the view model is pointless, as view models typically only have utility in their specific-ness towards a given view's purpose, but I find it useful because then we can do things like this:

```
protected GraphSubWindow(ISGViewModel viewModel)
{
    ViewModel = viewModel;
    m_ParentView = ViewModel.parentView;
    ParentView.Add(this);
    .....
}
```
Where setting up data and functionality for base window management stuff like binding references to a parent view and adding the view in question to the hierarchy etc, can all be handled by the `GraphSubWindow` base class, (which is the whole purpose for that class tbh).